src/sentry/event_manager.py

```diff
@@ -239,32 +293,105 @@ def __init__(self, data, version='5'):
         self.version = version
 
     def normalize(self):
-        # TODO(dcramer): store http.env.REMOTE_ADDR as user.ip
-        # First we pull out our top-level (non-data attr) kwargs
         data = self.data
+        errors = data.get('errors', [])
+
+        # Before validating with a schema, attempt to cast values to their desired types
+        # so that the schema doesn't have to take every type variation into account.
+        text = six.text_type
+        fingerprint_types = six.string_types + six.integer_types + (float, )
+
+        def to_values(v):
+            return {'values': v} if v and isinstance(v, (tuple, list)) else v
+
+        casts = {
+            'environment': lambda v: text(v) if v is not None else v,
+            'fingerprint': lambda v: map(text, v) if isinstance(v, list) and all(isinstance(f, fingerprint_types) for f in v) else v,
```

Comment:
not that it matters much because we do not really support python 3, but `map` on 3.x returns a generator.